### ISSUE TITLE
ENH: move numpy.distutils to numpy_distutils toplevel package

### ISSUE THREAD

**Issue by virtuald** [numpy/numpy#17620]
Cross-compiling scipy and other projects that depend on numpy's distutils is a huge pain right now, because to do it [in addition to lots of other details that you have to get right] you have to have both a native and cross-compiled version of numpy installed. It seems pretty unreasonable that I need a native version of numpy installed to compile scipy. One might ask, why is this needed?

Well, scipy's setup.py uses numpy's distutils fork for the fortran support (I think). Because numpy.distutils is a subpackage of numpy, if you're working with a cross-compiled version of numpy, it eventually tries to import some .so that wasn't compiled for your host and everything dies -- thus you have to have a native numpy installed to allow numpy.distutils to import correctly.

As far as I can tell, numpy's distutils fork is pure python and doesn't actually use anything else in numpy itself, and so is completely separable from numpy. If it were its own top-level package that scipy et al could use, then cross-compiling would be significantly less tricky.

The mechanics of this would work like so:

* contents of numpy.distutils would be moved to 'numpy_distutils' package
* importing numpy.distutils would raise a deprecation warning and redirect to numpy_distutils, probably using import hook magic
* scipy and other packages would now utilize numpy_distutils instead of numpy.distutils at build time

Initially, I considered proposing creating a separate pypi package for numpy_distutils, but I expect that would be more trouble than it's worth. One could also propose creating a new PEP 517 build backend, or move to cmake, or other huge changes, but those also seem more trouble than they're worth.

Thanks for your consideration.

**Comment by virtuald** [comment]
There were two positive responses to this on the mailing list, so I took a look at actually doing this tonight. My initial evaluation missed some things that would also need to change, here's the list so far:

* The `npy-pkg-config` functionality uses an import of numpy to find the files
  * This should be replaced with `importlib.util.find_spec` to avoid a direct import
* `numpy.distutils.misc_util` uses `numpy.compat`
* `get_numpy_include_dirs` in distutils calls `numpy.get_include`
  * This function probably should be moved to distutils?
* `numpy.distutils` calls things in `numpy.f2py`, which would also need to be moved to an independent toplevel package
  * Seems reasonable to move `numpy.f2py` to the `f2py` package?
* `numpy.f2py` would need to be modified to not call into numpy directly, luckily there are only two places this would be necessary
   * `numpy._pytester` is imported in `__init__` ... not sure of the ramifications of removing this. Maybe just replace it with None if the import fails?
   * f90mod_rules.py references `np.intp().itemsize`. I expect that could be hoisted out into some installed configuration file. Worst case scenario, it appears that `npy_intp` is the same as `Py_intptr_t`, which is the same as `distutils.sysconfig.get_config_vars('SIZEOF_UINTPTR_T')`... though I'm not sure that's public API
   * `numpy.f2py.rules` embeds the numpy version into files
   * `numpy.f2py.f2py_testing` uses `numpy.testing`
* The test infrastructure (`runtests.py`) expects everything to be under a single package hierarchy, and would need to be adapted to support these multiple toplevel modules.
* scipy uses numpy directly in some of it's building, so those elements need to be either removed or made accessible in numpy_distutils

I'm willing to do the work, but it does seem a bit more invasive than I had originally expected. I've pushed a draft PR with my initial work in #17632, so interested parties can see what kinds of changes I'm proposing. I don't really want to do any kind of finalization of the PR without some kind of 'yes we want this' response.

**Comment by rgommers** [comment]
> I'm willing to do the work, but it does seem a bit more invasive than I had originally expected. I've pushed a draft PR with my initial work in #17632, so interested parties can see what kinds of changes I'm proposing. I don't really want to do any kind of finalization of the PR without some kind of 'yes we want this' response.

Yes, this will be quite invasive:) The approach in your PR probably isn't going to work well as-is. You cannot really move something up one level there without having to create separate PyPI packages, plus wheels, plus conda-forge packages, etc. So it'd be better to discuss the end state you're aiming for here before doing a lot of work moving things around.

A few thoughts:

- Creating a completely separate `numpy-distutils` package is something I had expected we'd have to do years ago, and in principle makes sense.
- Longer term we may move everything into setuptools (see https://github.com/pypa/setuptools/issues/2372), so this could be a temporary change; making it as minimal as possible would be good.
- I'd rather avoid a separate `f2py` package unless absolutely necessary.
- One package one repo I'd say, so this would require putting `numpy_distutils` into a new repo, and pulling it back into its old place as a git submodule (necessary for backwards compat).
- It seems to me like it'd be best to create the new repo without making any changes to NumPy itself, and then move SciPy to it in a PR as a proof-of-concept that this actually does solve cross-compilation issues.

**Comment by eric-wieser** [comment]
> You cannot really move something up one level there without having to create separate PyPI packages, plus wheels, plus conda-forge packages

Why is that true? There's nothing to stop `pip install numpy` installing two top-level module, `numpy` and `numpy_distutils`. Yes, things become harder if you need to _install_ one without the other, but it sounds like @virtuald really only cares about _importing_ one without _importing_ the other.

**Comment by rgommers** [comment]
> Why is that true?

Ah, _technically_ that's not true, it is possible to have multiple top-level modules. It's very rare though and is very likely to introduce bugs because packaging tools don't really get tested for this. I had forgotten that it's possible at all, so let me just change my statement to "that's quite unhealthy, let's not do that".

The tricky bit here seems to be making sure that `numpy.distutils` doesn't need anything from `numpy`, either directly or transitively via `f2py`. Dealing with that first seems like the better idea, rather than doing everything at once. 

Once that's achieved, the normal thing to do is to split it off in a separate package and then vendor that back in.

**Comment by rgommers** [comment]
Let me add one other reason: an update of a separate package, and the reason I had expected to have had to do this separation already, is that bugs can be fixed in `numpy_distutils` and released _without_ a new NumPy release. Right now, if there's a bug in `numpy.distutils`, consuming packages are simply stuck with that until the minimum NumPy version they want to support no longer have that bug. And they have no other option than monkeypatching/duplicating. For example, SciPy has its own copy of `system_info.py` now (https://github.com/scipy/scipy/blob/master/scipy/_build_utils/system_info.py) because it needed to drop Accelerate support while still supporting ~5 NumPy versions.

**Comment by eric-wieser** [comment]
> It's very rare though

`matplotlib` does it, and isn't exactly a niche package. `mpl_toolkits` and `matplotlib` are two packages it provides.

**Comment by wkschwartz** [comment]
> that's quite unhealthy, let's not do that

pywin32 does this with `.pth` files, placing all of its submodules on the top level. I hate it.  Try grepping for pywin32 imports and miss `import win32file, pywintypes`, the former of which is the same as `import win32.lib.win32file` but the latter of which cannot be spelled differently despite the Python module appearing in `win32\lib\pywintypes.py`. (This highlights a second problem: top level packages with a different name from the PyPI project: you can’t `pip install win32`, much less `pip install win32file`).

I’d be sorry to see numpy go down this path.

**Comment by virtuald** [comment]
Sorry, been busy. Some responses:

> You cannot really move something up one level there without having to create separate PyPI packages, plus wheels, plus conda-forge packages, etc.

As stated above, you can. And my PR does this, and it seems to work. :)

> One package one repo I'd say, so this would require putting numpy_distutils into a new repo, 

In principle, I agree it would be better for `numpy.distutils` to be a separate pypi package -- but as I stated in the issue description, I think that's (at the moment) more trouble than it's worth. I think moving it to a separate toplevel package that's distributed with numpy is a better first step that will cause less compatibility issues.

> and pulling it back into its old place as a git submodule (necessary for backwards compat).

This isn't necessary either. In my PR I replaced numpy.distutils and numpy.f2py with a stub that utilizes modern python import hooks to redirects imports of the old package to the new package. A standalone version of the stub experiment is here: https://github.com/virtuald/transparent-pypkg-redir


> it sounds like @virtuald really only cares about importing one without importing the other.

> The tricky bit here seems to be making sure that numpy.distutils doesn't need anything from numpy, either directly or transitively via f2py. Dealing with that first seems like the better idea, rather than doing everything at once.

Yes, that is exactly the thing that I care about. The PR actually fixes those issues with the exception of the f2py problem. They could be separated out into separate initial PRs if that would be easier to review -- though note that they're not that useful without moving the package.

Unfortunately, you can't make it so that numpy.distutils doesn't require things from f2py without breaking backwards compat. It looks like numpy.distutils can be used to generate source files via f2py?

https://github.com/numpy/numpy/blob/b7c27bd2a3817f59c84b004b87bba5db57d9a9b0/numpy/distutils/command/build_src.py#L530

> pywin32 does this with .pth files, placing all of its submodules on the top level. I hate it. [snip] I’d be sorry to see numpy go down this path.

Because (as far as I can tell) `numpy.distutils` and `numpy.f2py` are primarily for users building things as opposed to code that users interact with and write on a daily basis, I think it's much less bad to have them in separate packages.

**Comment by virtuald** [comment]
It's worth restating that the primary problem that needs to be solved to make cross-compiling work is that when I do an 'import numpy' it (rightly) imports a bunch of compiled C extensions and everything breaks. An alternative approach to make cross compiling work is to just have that break gracefully.. but that seems like it could be potentially very fragile.

A terrible approach that occurs to me is that you could have a second dummy package called `numpy_build_mode` that someone doing numpy-dependent builds imports before importing `numpy`. Then when the C extension imports fail, it checks sys.modules for the presence of that module and suppresses any errors (instead of the giant error messages you would normally get). 

That's really terrible, but would be much simpler in terms of backwards compatibility.

**Comment by dazza-codes** [comment]
Watching this, because reducing the numpy package size for production purposes is useful for AWS lambda layer packaging and it seems like this issue might help with that.  A similar approach might be useful for `numpy.testing`.  A similar request is open on pandas
- https://github.com/pandas-dev/pandas/issues/30741

**Comment by virtuald** [comment]
`numpy.testing` does not effect cross-compilation, so I don't care about it. :)

**Comment by rgommers** [comment]
> In principle, I agree it would be better for `numpy.distutils` to be a separate pypi package -- but as I stated in the issue description, I think that's (at the moment) more trouble than it's worth.

It'd be pure Python, so it really isn't all that hard.

> I think moving it to a separate toplevel package that's distributed with numpy is a better first step that will cause less compatibility issues.

I'm not sure why you think that, even for the sole purpose of cross-compiling SciPy. How do you imagine making that work? If it's a standalone package we can discuss SciPy relying on it. If it's part of NumPy, then that becomes much harder, given that SciPy needs to support 4-5 NumPy versions.

**Comment by rgommers** [comment]
> Yes, that is exactly the thing that I care about. The PR actually fixes those issues with the exception of the f2py problem. They could be separated out into separate initial PRs if that would be easier to review -- though note that they're not that useful without moving the package.

Multiple smaller PRs would indeed be a lot easier to review I think.

**Comment by rgommers** [comment]
> Watching this, because reducing the numpy package size for production purposes is useful for AWS lambda layer packaging and it seems like this issue might help with that. A similar approach might be useful for `numpy.testing`. A similar request is open on pandas

That is not a good idea for multiple reasons, from diagnosing the many bug reports we receive to it not being optimal for AWS Lambda in the first place (you'd want to, e.g., strip docstrings and other things you don't care about).

There is some tooling for AWS Lambda floating around to get minimal binaries. That really should be consolidated somewhere by someone who cares about this use case. I've done this before with a Docker workflow to be able to squeeze NumPy + SciPy + Pandas + application code into 125 MB. That's really a different issue from cross-compiling, so let's not mix it into this issue.

**Comment by virtuald** [comment]
I've split out some of the giant PR into several little PRs (#17728, #17729, #17730, #17731, #17732) that can be reviewed now once the CI passes. I think just about everything else in the PR are directly related to moving f2py/distutils and so they don't really stand alone.

**Comment by jaskij** [comment]
I only know basics of Python, but can offer some testing under Yocto/OpenEmbedded if it would help.

**Comment by virtuald** [comment]
I'm really sorry, but my scheduled got filled up so this set of PRs still has some loose ends. Might be able to do it later this month, I'm still interested in the topic.

**Comment by jaskij** [comment]
We've all got our lifes, no problem. I found this series after 12+ hours fighting with SciPy and Yocto and just couldn't pass by. It's great that you haven't abandoned it.

As an additional question - any chance of backporting this to 1.17? That's what's included in Yocto's current LTS.

**Comment by eric-wieser** [comment]
> As an additional question - any chance of backporting this to 1.17?

Numpy only maintains two versions at a time - so the only versions this patch could reach are 1.19 onwards, and it's likely that this won't make 1.19 at all.

**Comment by pearu** [comment]
My first reaction to this enhancement was that this is a bad idea, mostly because of YAGNI and being Pandora's box of bugs.

**OT Warning START**: my interpretation of this enhancement includes a possible longer-term wish to maintain numpy, numpy_distutils, f2py, and numpy_testing as separate projects.

In general, whenever a package is split into many interoperable packages, one increases a maintenance burden. For instance, as noted above, scipy has to support several versions of numpy. Similarly, when numpy is split into numpy_distutils (btw, in past we had scipy_distutils), f2py, and perhaps numpy_testing, then in long term one needs to maintain the support of:
- numpy for multiple versions of numpy_distutils and numpy_testing
- f2py for multiple versions of numpy_distutils, numpy, numpy_testing
- numpy_distutils for multiple versions of f2py

I think the most difficult combinations would be
- f2py dependence on numpy and numpy_distutils, and
- numpy dependence on numpy_distutils (the source code generation part)

On top of the above, add scipy that will need to maintain support for multiple versions of numpy, f2py, numpy_distutils, numpy_testing.

On top of the above 2, add burden from newly introduced bugs when implementing the package split.

If the aim is to distribute/package/import/.. numpy without numpy_distutils, f2py, and numpy_testing in resource tight environments, then there is an alternative approach: split numpy into numpy_core and numpy^2 where
- numpy_core would contain everything in current numpy except numpy_distutils, f2py, numpy_testing 
- and numpy^2 would be the current numpy except importing numpy_core so that the numpy backward compatibility is preserved.

If something needs to be split, then it would be scipy, imho. 

**OT Warning END**

Since the aim of this issue is to make scipy cross-compilation easier, I am not sure it is worth effort considering the invasive nature of the PR and a high chance of breaking a working system in a nontrivial way with no advantage to the end-users of numpy and scipy.

If it will be agreed to purchase resolving this issue as described, I would recommend implementing it slowly:
- make a long term goal of the split
- introduce new packages and gradually start moving code from numpy to new packages while maintaining backward compatibility
- give time for downstream software (scipy may not be the only affected project) to adjust to the new package structure (numpy will need to trigger appropriate deprecation warnings, etc).

My 2 cents.

**Comment by pearu** [comment]
> Well, scipy's setup.py uses numpy's distutils fork for the fortran support (I think). Because numpy.distutils is a subpackage of numpy, if you're working with a cross-compiled version of numpy, it eventually tries to import some .so that wasn't compiled for your host and everything dies -- thus you have to have a native numpy installed to allow numpy.distutils to import correctly.

To avoid importing .so modules, I wonder if using the `builtins.__NUMPY_SETUP__ = True` hook would work around this problem? Or at least use the same idea? For instance, consider
```
$ python
>>> import numpy
>>> numpy.core._multiarray_umath
<module 'numpy.core._multiarray_umath' from '/home/pearu/git/pearu/numpy/numpy/core/_multiarray_umath.cpython-39-x86_64-linux-gnu.so'>
>>> ^D
$ python
>>> import builtins
>>> builtins.__NUMPY_SETUP__ = True
>>> import numpy
Running from numpy source directory.
>>> numpy.core._multiarray_umath
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'numpy' has no attribute 'core'
```

**Comment by pearu** [comment]
To continue with the above idea, consider
```
$ python
Python 3.9.1 | packaged by conda-forge | (default, Jan 10 2021, 02:55:42) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import builtins
>>> builtins.__NUMPY_SETUP__ = True
>>> import numpy
Running from numpy source directory.
>>> import numpy.distutils
>>> import numpy.f2py
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/pearu/git/pearu/numpy/numpy/f2py/__init__.py", line 11, in <module>
    from . import f2py2e
  File "/home/pearu/git/pearu/numpy/numpy/f2py/f2py2e.py", line 23, in <module>
    from . import rules
  File "/home/pearu/git/pearu/numpy/numpy/f2py/rules.py", line 83, in <module>
    from . import f90mod_rules
  File "/home/pearu/git/pearu/numpy/numpy/f2py/f90mod_rules.py", line 63, in <module>
    if ((.not.allocated(d)).and.(s(1).ge.1)) then""" % np.intp().itemsize
AttributeError: module 'numpy' has no attribute 'intp'
```
which means that while `numpy.distutils` will be readily available for cross-compilation of scipy but numpy.f2py requires some work to remove its dependence on numpy.core and there must be a way to specify the target-specific parameters such as the `intp.itemsize`. I think this approach will be considerably less invasive than the issue proposal.

**Comment by jaskij** [comment]
> If the aim is to distribute/package/import/.. numpy without numpy_distutils, f2py, and numpy_testing in resource tight environments, then there is an alternative approach: split numpy into numpy_core and numpy^2 where

The aim here is the opposite: to have numpy_distutils without importing the whole of numpy, to make cross-builds (ie OpenEmbedded is my use case) of scipy and similar not depend on native build of numpy. It seems that numpy_distutils can be a pure Python module, greatly simplifying things. As @virtuald wrote, numpy has binary dependencies which can be problematic in such an environment.

To reiterate: this is not about what is happening on the target device, but cross-building the image on a separate host.

**Comment by pearu** [comment]
Thanks, @jaskij for the explanation. Here's a 4-line implementation of numpy_distutils.py module that is pure Python and should provide the required functionality:
```
# numpy_distutils.py
import builtins as _builtins 
_builtins.__NUMPY_SETUP__ = True
from numpy.distutils import *
_builtins.__NUMPY_SETUP__ = False
```

However, as noted above, this will not be sufficient to build scipy: numpy.f2py has to be adjusted to not use numpy.core tools.

**Comment by jaskij** [comment]
@pearu to be honest, I didn't understood OT meant off-topic the first time and stopped reading after the paragraph I quoted from, my bad.

OpenEmbedded, the system I'm using (and which is the Linux Foundation-backed industry standard), is so good about separating host and  target and setting up correct variables I managed to get it building, with the help of [gpanders/meta-scipy](https://github.com/gpanders/meta-scipy).

There are still some issues which complicate my solution, but at least for this one environment there's a working workaround.

The way OpenEmbedded works, ultimately, the perfect solution would be the ability to _package_ numpy_distutils and f2py without numpy itself. The system works around packages, whether it's build dependencies for the cross host or runtime dependencies.

Taking a closer look at gpanders' [python3-scipy_1.5.3.bb](https://github.com/gpanders/meta-scipy/blob/master/recipes-devtools/python/python3-scipy_1.5.3.bb), which is the instruction for OpenEmbedded how to build scipy, you can see the dependencies list in [line 12](https://github.com/gpanders/meta-scipy/blob/master/recipes-devtools/python/python3-scipy_1.5.3.bb#L12). Without any prior knowledge, having a package depend on both numpy (which is the target numpy) and numpy-native (which is numpy built for host) is a code smell.

**Comment by pearu** [comment]
@jaskij , no problem, I realize my "this is bad idea" comment may have sound harsh but I meant it only in a constructive way :)

Another way to make `numpy.{distutils, f2py}` easier for OpenEmbedded to handle is to have two setup.py scripts (actually, one is sufficient but its functionality could be tweaked by some environment variable):
- `setup.py` that packages/installs numpy as it does now,
- `NUMPY_SETUP_NO_CORE=1 setup.py` that packages/installs numpy as, say, `numpy_nocore` that includes only numpy.distutils and numpy.f2py and would be made available to OpenEmbedded like systems. Notice that `NUMPY_SETUP_NO_CORE=1` would change only the parent package name (`numpy` -> `numpy_nocore`) and it would discard subpackages such as `numpy.core` and others when installing/packaging.

This approach could be implemented with minimal changes to numpy and would be an extension of numpy (introduces a new package) rather than a major rewrite of a numpy packaging structure that #17632 is, IMHO.

**Comment by jaskij** [comment]
> NUMPY_SETUP_NO_CORE=1 setup.py that packages/installs numpy as, say, numpy_nocore that includes only numpy.distutils and numpy.f2py and would be made available to OpenEmbedded like systems. Notice that NUMPY_SETUP_NO_CORE=1 would change only the parent package name (numpy -> numpy_nocore) and it would discard subpackages such as numpy.core and others when installing/packaging.

This would work in my environment, but I am not sure about other cross-compilation systems and environments (while industry standard in industrial, OpenEmbedded is by far not the only system). The only issue with this approach would be that then scipy's (and other) setup.py would also need to know to import numpy_nocore, but that probably is easily patched downstream (I'm thinking about current and older versions which probably would not include this upstream).

Another issue is numpy distutils being somewhat greedy with paths, as seen [in this patch](https://github.com/gpanders/meta-scipy/blob/master/recipes-devtools/python-numpy/python3-numpy/0001-Disable-runtime_lib_dirs-for-cross-compile.patch) - the default behavior is probably good for normal users, but an option to disable it _without_ patching would be preferred in cross environments.

**Comment by HaoZeke** [comment]
Is anyone against closing this and the related #17632 for now? `distutils` is on its last legs as is; and also a change like this should probably be accompanied by some discussion on the mailing list.

**Comment by rgommers** [comment]
I agree with closing this PR. There is interesting info here though - good cross-compilation support will be one of the most important benefits of moving SciPy to Meson (see [blog post](https://labs.quansight.org/blog/2021/07/moving-scipy-to-meson/)). We will be merging support into SciPy master within weeks I expect. It'd be great to test that with OpenEmbedded (and any other cross-compilation setup people here are interested in) to ensure everything works as expected. 

What I'd like to do is open a separate SciPy issue for cross-compilation setups and Cc people working on those there, so once the support is in place we have a single tracking issue to collect use cases. There are more mentioned in https://news.ycombinator.com/item?id=27953673.

There may still be a benefit from decoupling `numpy.f2py` from `numpy.distutils` (although it's small, things should "just work" because the f2py output is not platform-dependent), and that should be done - just not in such an invasive way.

**Comment by HaoZeke** [comment]
Closing by unanimous agreement.
